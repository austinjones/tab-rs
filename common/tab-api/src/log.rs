//! A global, configurable level filters
//! The command, daemon, and pty honor this level on startup.
//!
//! The main executible configures if the --log <level> option is provided
//!
//! get_level() returns None unless set_level has been called.
use std::sync::atomic::{AtomicU8, Ordering};

use log::LevelFilter;

// info level
static LOG_LEVEL: AtomicU8 = AtomicU8::new(0);

pub fn set_level(level: LevelFilter) {
    let byte_repr = discriminant_of(level);
    LOG_LEVEL.store(byte_repr, Ordering::SeqCst);
}

pub fn get_level() -> Option<LevelFilter> {
    level_of(LOG_LEVEL.load(Ordering::SeqCst))
}

pub fn get_level_str() -> Option<&'static str> {
    let level = get_level();

    if let None = level {
        return None;
    }

    match level.unwrap() {
        LevelFilter::Off => Some("off"),
        LevelFilter::Error => Some("error"),
        LevelFilter::Warn => Some("warn"),
        LevelFilter::Info => Some("info"),
        LevelFilter::Debug => Some("debug"),
        LevelFilter::Trace => Some("trace"),
    }
}

// LevelFilter has a from_usize method, but it's private
// we have to redo the cases here, and support None as 0
fn discriminant_of(filter: LevelFilter) -> u8 {
    match filter {
        LevelFilter::Trace => 1,
        LevelFilter::Debug => 2,
        LevelFilter::Info => 3,
        LevelFilter::Warn => 4,
        LevelFilter::Error => 5,
        LevelFilter::Off => 6,
    }
}

fn level_of(filter: u8) -> Option<LevelFilter> {
    match filter {
        0 => None,
        1 => Some(LevelFilter::Trace),
        2 => Some(LevelFilter::Debug),
        3 => Some(LevelFilter::Info),
        4 => Some(LevelFilter::Warn),
        5 => Some(LevelFilter::Error),
        6 => Some(LevelFilter::Off),
        _ => unreachable!("unreachable level discriminant"),
    }
}
