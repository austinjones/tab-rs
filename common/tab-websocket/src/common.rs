use crate::WebSocket;
use futures::SinkExt;
use log::error;
use serde::Serialize;
use tungstenite::{Error, Message};

pub async fn send_message<Msg: Serialize>(websocket: &mut WebSocket, message: Msg) {
    let encoded = bincode::serialize(&message);

    if let Err(e) = encoded {
        error!("failed to encode message: {}", e);
        return;
    }

    let sent = websocket.send(Message::Binary(encoded.unwrap())).await;
    if let Err(e) = sent {
        error!("failed to send message: {}", e);
    }
}

pub async fn send_close(websocket: &mut WebSocket) {
    if let Err(e) = websocket.send(Message::Close(None)).await {
        match e {
            Error::ConnectionClosed | Error::AlreadyClosed => {
                return;
            }
            _ => {
                error!("failed to send close frame: {}", e);
            }
        }
    }
}
