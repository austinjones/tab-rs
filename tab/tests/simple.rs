use insta::assert_snapshot;

mod common;
use common::*;
use log::info;

/// Tests that a session can be established, disconnected from, and re-established
/// Covers connection, ctrl-W, disconnection, and scrollback
#[tokio::test]
async fn reconnect() -> anyhow::Result<()> {
    let mut session = TestSession::new()?;

    let result = session
        .command()
        .tab("simple/")
        .await_stdout("$", 5000)
        .stdin("exit\n")
        .await_stdout("exit", 200)
        .complete_snapshot()
        .run()
        .await?;

    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("after", &result.snapshot);
    info!("after: {}", result.snapshot);

    Ok(())
}
