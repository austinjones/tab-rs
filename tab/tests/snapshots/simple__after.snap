---
source: tab/tests/simple.rs
expression: "&result.snapshot"
---
$ exit

